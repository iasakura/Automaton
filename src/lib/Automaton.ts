import { DataSet, Network, Options } from 'vis-network/standalone';
import jsyaml from 'js-yaml';

export type IdType = string;
export type Alphabet = string & { length: 1 };

class State {
  next: Map<Alphabet, IdType>;
  constructor(public id: IdType) {
    this.next = new Map();
  }

  push(char: Alphabet, state: IdType) {
    this.next.set(char, state);
  }

  getNext(alpha: Alphabet): IdType | undefined {
    return this.next.get(alpha);
  }
}

export class Automaton {
  constructor(
    public init_state: IdType,
    private states: Map<IdType, State>,
    private accepting_state: Set<IdType>
  ) {}

  getNext(cur: IdType, input: Alphabet): IdType | undefined {
    return this.states.get(cur)?.getNext(input);
  }

  getTransition(cur: IdType): Map<Alphabet, IdType> | undefined {
    return this.states.get(cur)?.next;
  }

  isAcceptingState(state: IdType) {
    return this.accepting_state.has(state);
  }

  static parse(text: string): Automaton {
    let states = new Map<IdType, State>();

    const yaml = jsyaml.load(text) as {
      init_state: Object;
      nodes: Object[];
      transitions: Record<string, Record<string, Object>>;
      accepting_state: Object[];
    };

    if (yaml.init_state === undefined) {
      throw Error('Cannot find init_state');
    }
    const init_state = yaml.init_state.toString();

    if (yaml.accepting_state === undefined) {
      throw Error('Cannot find accepting_state');
    }
    const accepting_state = new Set(
      yaml.accepting_state.map((s) => s.toString())
    );

    for (const state of yaml.nodes) {
      const s = state.toString();
      states.set(s, new State(s));
    }

    for (const [from, alpha_to] of Object.entries(yaml.transitions)) {
      for (const [alpha, to] of Object.entries(alpha_to)) {
        if (alpha.length !== 1) {
          throw Error('The length of alpha should be 1');
        }
        if (!states.has(from)) {
          throw Error(`The state ${from} doesn't exists`);
        }
        states.get(from)?.push(alpha as Alphabet, to.toString());
      }
    }

    return new Automaton(init_state, states, accepting_state);
  }

  asVisNetwork(container: HTMLElement, options: Options): Network {
    const nodes = Array.from(this.states.keys()).map((id) => {
      return { id: id, label: id };
    });

    let edges: { from: string; to: string; label: string }[] = [];
    this.states.forEach((state, from) => {
      state.next.forEach((next, alpha) => {
        edges.push({ from, to: next, label: alpha });
      });
    });

    return new Network(
      container,
      {
        nodes: new DataSet(nodes),
        edges: new DataSet(edges),
      },
      options
    );
  }
}
