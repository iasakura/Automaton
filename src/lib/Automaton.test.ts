import { Automaton } from './Automaton';

test('has even a', () => {
  // let states = '0 1';
  let transition = `\
init_state: 0
nodes:
  - 0
  - 1
transitions:
  0:
    a: 1
    b: 0
  1:
    a: 0
    b: 1
accepting_state:
  - 0
`;

  const automaton = Automaton.parse(transition);

  console.log(JSON.stringify(automaton));

  expect(automaton).toMatchSnapshot();
});
