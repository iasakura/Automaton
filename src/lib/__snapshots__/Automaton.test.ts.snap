// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has even a 1`] = `
Automaton {
  "accepting_state": Set {
    "0",
  },
  "init_state": "0",
  "states": Map {
    "0" => State {
      "id": "0",
      "next": Map {
        "a" => "1",
        "b" => "0",
      },
    },
    "1" => State {
      "id": "1",
      "next": Map {
        "a" => "0",
        "b" => "1",
      },
    },
  },
}
`;
